Add DataPostprocessors::BoundaryIds.

bangerth opened the pull request:
I meant to do that for a long time already, having written this code for another project. An email on the list today reminded me of it.

/rebuild

tjhei commented on doc/doxygen/headers/glossary.h:
```diff
@@ -346,6 +346,43 @@
+ * they simply ignore this information and only output information obtained
+ * from the mesh (namely, the boundary indicators).
+ *
+ * The example code above uses FE_Q as the finite element. This is appropriate
```
Why not use a DGQ0?

bangerth replied:
Sure, I could do that as well. In principle, I could probably even do `FE_Nothing`. Want me to change this?

tjhei replied:
That is also an option. I guess it really doesn't matter. :-)

tjhei commented on source/numerics/data_postprocessor.cc:
```diff
@@ -196,6 +196,36 @@ DataPostprocessorTensor<dim>::get_needed_update_flags() const
+    for (auto &output : computed_quantities)
+      {
+        AssertDimension(output.size(), 1);
+        output(0) = cell->face(face)->boundary_id();
```
What happens if this is run with ``DataOutFaces(/*surface_only=*/ false)``? Does this call fail with an assert? If not, can we document what values will be set in the interior?

bangerth replied:
Ah, good question. 

I could go either way. My inclination is that something that only lives on the surface should only be used on the surface, and throw an assertion otherwise. Would you agree? If so, I'll implement that.

tjhei replied:
I would think it is okay to run with internal faces and output -1 in the interior (documenting that would be great).
I imagine a situation where someone has other data to visualize in the interior like manifold ids.

tjhei reviewed (commented):
Very nice. I have a few questions.

bangerth commented:
I've added a couple of commits that now let us deal with internal faces as well.

tjhei reviewed (approved):
very nice, thank you for the improvements!

kronbichler commented on doc/doxygen/headers/glossary.h:
```diff
@@ -346,6 +346,43 @@
+ *    DataOutFaces<dim> data_out_faces;
+ *    FE_Q<dim>         dummy_fe(1);
+ *
+ *    DoFHandler<dim>   dummy_dof_handler(triangulation);
+ *    dummy_dof_handler.distribute_dofs(dummy_fe);
+ *
+ *    Vector<double> dummy_solution (dummy_dof_handler.n_dofs());
+ *
+ *    data_out_faces.attach_dof_handler(dummy_dof_handler);
+ *    data_out_faces.add_data_vector(dummy_solution, boundary_ids);
```
I have not tried this myself, but can't you just set
```c++
  data_out_faces.attach_triangulation(triangulation);
  data_out_faces.build_patches();
```
It might be that you still need to pass a dummy vector (of cell values). At least for `DataOut`, we worked towards not needing to set a `DoFHandler` in case we only write cell data, see also the documentation of the function:
https://github.com/dealii/dealii/blob/b1b07cec07e64dcb0424ae809ba8cf3b3dd86560/include/deal.II/numerics/data_out_dof_data.h#L654-L664
And in case this does not work, I think it should be a rather limited step to actually remove the need for `DoFHandler` also in `DataOutFaces`, and what I would prefer over teaching to create dummy values.

tjhei replied:
I am not 100% but I think DataOut also requires a DoFHandler if you want to use a DataPostProcessor.

bangerth replied:
@tjhei is right: When using a postprocessor, we need a `dof_handler` object:
```
template <int dim, int patch_dim, int spacedim, int patch_spacedim>
template <typename VectorType>
void
DataOut_DoFData<dim, patch_dim, spacedim, patch_spacedim>::add_data_vector(
  const VectorType &                 vec,
  const DataPostprocessor<spacedim> &data_postprocessor)
{
  Assert(dofs != nullptr,
         Exceptions::DataOutImplementation::ExcNoDoFHandlerSelected());
  add_data_vector(*dofs, vec, data_postprocessor);
}
```
So that won't work.

bangerth commented:
Is there anything else that y'all want me to do? I've added another comment about the choice of element, and I've added a test that verifies that this also works with `FE_Nothing`.
